Backwards compatible control types refactor

seanbudd opened the pull request:
### Link to issue number:

Closes #10732

### Summary of the issue:

`controlTypes.py` is a large file with many constants. To maintain backwards compatibility, the file length would increase to 1000+ LOC. It would be an improvement to split this code up, and group the constants into enums. As 2021.2 is not a compatibility breaking release, all changes must be backwards compatible. 

### Description of how this pull request fixes the issue:

This is a large refactor. The changes are split into individual commits for reviewers.

The first commit can be reviewed using the following tool

```sh
for file in $(ls source/controlTypes/*.py)
do
git diff master:source/controlTypes.py 8642581:$file > $file-controlTypes.diff
done
```

The second commit is generated using the following sed commands. 

```bash
sed -i'' -bre "s/ROLE_([A-z0-9_]+) ?= ?/\t\1 = /g" source/controlTypes.py
sed -i'' -bre "s/ROLE_([A-z0-9_]+) ?:([^ ])/Role.\1: \2/g" source/controlTypes.py
sed -i'' -bre "s/ROLE_([A-z0-9_]+)/Role.\1/g" source/controlTypes.py

sed -i'' -bre "s/STATE_([A-z0-9_]+) ?= ?/\t\1 = /g" source/controlTypes.py
sed -i'' -bre "s/STATE_([A-z0-9_]+) ?:([^ ])/State.\1: \2/g" source/controlTypes.py
sed -i'' -bre "s/STATE_([A-z0-9_]+)/State.\1/g" source/controlTypes.py
```

The mixin strategy follows what is outlined for python 3.7 [here](https://archive.is/IfNLZ). Mixins with Enums are fragile, and this code will have to be changed when updating (or downgrading python) Python.

### Testing strategy:

Manual testing. 

There is unit testing for the controlTypes module which should help ensure that compatibility doesn't break.

This module is fairly widely utilised, so any breaking changes are likely to be picked up quickly.

### Known issues with pull request:

Changes can't be tracked backwards using git blame to `controlTypes.py` due to how the files are split up. As such, I've linted these files.

I haven't implemented the enums in the wider NVDA codebase. When we break the backwards compatibility, by updating to 2022.1, we will need to update the rest of the NVDA code beforehand (see #12549).

### Change log entries:

For developers:
- `controlTypes` has been split up into various submodules. `ROLE_*` and `STATE_*` constants should be replaced to their equivalent `Role.*` and `State.*` before 2022.1. (#12510)
- `roleLabels`, `stateLabels` and `negativeStateLabels` have been deprecated, usages such as `roleLabels[ROLE_*]` should be replace to their equivalent `Role.*.displayString` or `State.*.negativeDisplayString` before 2022.1. (#12510)

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Pull Request description is up to date.
- [x] Unit tests.
- [x] System (end to end) tests.
- [x] Manual testing.
- [x] User Documentation.
- [x] Change log entry.
- [x] Context sensitive help for GUI changes.
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers

leonardder commented:
This is an awesome change!

Could you consider the following while at it?

1. Create an abstract class that defines the displayString property e.g. `EnumWithTranslatableDisplayString`
2. Base the following classes on this abstract class: IsCurrent, Role and State
3. Role and State's displayString should return the role/state label, respectively
4. Instead of letting code use the role and state label dictionaries, use the displayString property instead.

leonardder commented on source/controlTypes/role.py:
```diff
@@ -0,0 +1,521 @@
+from typing import Dict, Set
+
+
+class ROLE(IntEnum):
```
```suggestion
class Role(IntEnum):
```

leonardder commented on source/controlTypes/state.py:
```diff
@@ -0,0 +1,156 @@
+from typing import Dict
+
+
+class STATE(IntEnum):
```
```suggestion
class State(IntEnum):
```

seanbudd commented:
> This is an awesome change!
> 
> Could you consider the following while at it?
> 
> 1. Create an abstract class that defines the displayString property e.g. `EnumWithTranslatableDisplayString`
> 2. Base the following classes on this abstract class: IsCurrent, Role and State
> 3. Role and State's displayString should return the role/state label, respectively
> 4. Instead of letting code use the role and state label dictionaries, use the displayString property instead.

I think this would be suited better as a mixin. I've done a demo as follows, pending thoughts on yourself and other reviewers. 

```diff
diff --git a/source/controlTypes/isCurrent.py b/source/controlTypes/isCurrent.py
index 1d0744e98..506d24a30 100644
--- a/source/controlTypes/isCurrent.py
+++ b/source/controlTypes/isCurrent.py
@@ -3,13 +3,35 @@
 # See the file COPYING for more details.
 # Copyright (C) 2007-2021 NV Access Limited, Babbage B.V.
 
+from abc import ABCMeta, abstractproperty
 from enum import Enum
 from typing import Dict
 
 from logHandler import log
 
 
-class IsCurrent(Enum):
+class DisplayStringEnumMixin(ABCMeta):
+	@abstractproperty
+	def displayStringLabels(self) -> Dict[Enum, str]:
+		pass
+
+	@abstractproperty
+	def defaultValue(self) -> Enum:
+		pass
+
+	@property
+	def displayString(self: Enum) -> str:
+		"""
+		@return: The translated UI display string that should be used for this value of the enum
+		"""
+		try:
+			return self.displayStringLabels[self]
+		except KeyError:
+			log.debugWarning(f"No translation mapping for: {self}")
+			return self.displayStringLabels[self.defaultValue]
+
+
+class IsCurrent(Enum, DisplayStringEnumMixin):
 	"""Values to use within NVDA to denote 'current' values.
 	These describe if an item is the current item within a particular kind of selection.
 	EG aria-current
@@ -23,17 +45,12 @@ class IsCurrent(Enum):
 	TIME = "time"
 
 	@property
-	def displayString(self):
-		"""
-		@return: The translated UI display string that should be used for this value of the IsCurrent enum
-		"""
-		try:
-			return _isCurrentLabels[self]
-		except KeyError:
-			log.debugWarning(f"No translation mapping for: {self}")
-			# there is a value for 'current' but NVDA hasn't learned about it yet,
-			# at least describe in the general sense that this item is 'current'
-			return _isCurrentLabels[IsCurrent.YES]
+	def displayStringLabels(self):
+		return _isCurrentLabels
+
+	@property
+	def defaultValue(self):
+		return self.YES
```

leonardder commented:
> I think this would be suited better as a mixin. 

Ugh, I actually meant to say that 😳

> I've done a demo as follows, pending thoughts on yourself and other reviewers.

This is perfect! I'd probably make _displayStringLabels pseudo private to the enum, prefixing it with an underscore.

seanbudd commented:
I created the mixin and renamed `ROLE` and `STATE` to be UpperCamelCase. I wasn't sure where it made sense to put said mixin though, so a new module was created. Any suggestions on where it may be better suited, or a rename of the new module structure, would be appreciated.

leonardder commented on source/utils/mixins.py:
```diff
@@ -0,0 +1,58 @@
+class DisplayStringEnumMixin(ABC):
+	"""
+	This mixin can be used with a class which subclasses Enum to provided translated display strings for
+	members of the enum. The abstract properties must be overriden.
```
```suggestion
	members of the enum. The abstract properties must be overridden.
```

leonardder commented on source/utils/mixins.py:
```diff
@@ -0,0 +1,58 @@
+		try:
+			return self._displayStringLabels[self]
+		except KeyError:
+			log.debugWarning(f"No translation mapping for: {self}")
```
I think it will be easy to miss these when on debugWarning.

```suggestion
			log.error(f"No translation mapping for: {self}")
```

leonardder commented on source/controlTypes/__init__.py:
```diff
@@ -0,0 +1,236 @@
+
+# Added to maintain backwards compatibility, marked for deprecation to be removed in 2022.1
+# usages to be replaced by Role.*.displayString and State.*.displayString
+if version_year < 2022:
```
I think we should be careful to introduce automatic deprecation like this. It would be better to remove the code in 2022 anyway. Furthermore, in the current state in core, this would just break as soon as the version is updated to 2022.x.

seanbudd replied:
The code will be removed in 2022. Preparing for code removal in 2022 is the sole purpose of this PR, if we need to delay the deprecation we can simply adjust the `if` condition with a later version.

lukaszgo1 replied:
For what its worth I do share @leonardder 's concerns here.  With this change it is extremely easy to remove parts of the code unintentionally. Normally when the new development cycle begins update of `version_year` is one of the first things done and then we can deprecate whatever needs to be deprecated piece by piece. Making backward compatibility code dependent on the value  of `version_year` forces us to one of the following:
- Deprecate everything at the beginning of the 2022.1 dev cycle in one  piece  - making it unnecessarily harder to detect regressions
- Wait with the `version_year` upgrade until all deprecation's are done - while it is not problematic as such this seems to be rather unnecessary.

If we really need to make this back compat stuff automatic then I would suggest to use `BACK_COMPAT_TO` from the `addonAPIVersion` as it is incremented just after first beta is released.

seanbudd replied:
Ah, you are right `BACK_COMPAT_TO` would definitely be the better option here. I was not thinking about when exactly we would be updating `version_year`.

seanbudd replied:
It's less about being automatic, more about being self-documenting. This code should be removed or fixed before it becomes an issue.

feerrenrut replied:
I think there is a little confusion here. It is being marked as deprecated (_"deprecation is the discouragement of use of some terminology, feature, design, or practice, typically because it has been superseded or is no longer considered efficient or safe, without completely removing it or prohibiting its use")_ as of 2021.2. The goal is to remove it and prohibit it's use in the 2022.1 release. The earlier in the release we can make remove access to deprecated code, the earlier addon are aware of the change. With this mechanism, the code is effectively removed at the start of the dev cycle, as soon as we update the version number. When code is deprecated, addons should start the process of removing dependence on it. This mechanism allows addon's to test that they don't rely on the deprecations, by manually updating the version number used in NVDA.

When we get to 2022.1 we would simply remove this code.

This mechanism does mean that all core code needs to be updated to use the replacement approach. I'd say that was true anyway.

> If we really need to make this back compat stuff automatic then I would suggest to use BACK_COMPAT_TO from the addonAPIVersion as it is incremented just before first beta is released.

That is when we commit to the API, we want to have been "living with" the new API for the duration of the release dev cycle first so that any issues can be discovered. Committing to the API removals should happen as early in the release dev cycle as possible. Yes, this results in no warning to uses of addons that didn't note deprecations in prior releases. This is to be expected when running alpha software.

lukaszgo1 replied:
As long as the entire NVDA's code base is updated to the enums *before* 2022.1 dev cycle begins and this is only to stop add-ons from accessing these old constants I'm fine with `version_year` being used here. For previous backward compatibility breaking releases we used to remove deprecate code at the same time when the code base was updated with the new approach and that prompted my comment.

feerrenrut replied:
>As long as the entire NVDA's code base is updated to the enums before 2022.1 dev cycle begins

@lukaszgo1 yes, we have a mechanism for this, a new issue with a milestone has been created. If for some reason that work can't happen, we have options:
- postpone updating the version number
- change these checks to a later version number

> For previous backward compatibility breaking releases we used to remove deprecate code at the same time when the code base was updated with the new approach and that prompted my comment.

For several other enum updates we have converted the internal usages of the deprecated constants at the same time as introducing the enum. For this PR, we won't do that immediately. NVDA will continue to use the constants for a short time. This will help to ensure that this change is actually backwards compatible.

Once the backwards compatibility is confirmed, we can go ahead with removing internal usages of the newly deprecated constants. Then finally in 2022.1 we can remove the deprecated code entirely.

feerrenrut commented on source/controlTypes/__init__.py:
```diff
@@ -0,0 +1,236 @@
+
+
+# Added to maintain backwards compatibility, marked for deprecation to be removed in 2022.1
+if BACK_COMPAT_TO[0] < 2022:
```
I think realizing the deprecations, and removing access to these constants should happen as early as possible in the release dev cycle. This isn't updated until the end. NVDA version should be used instead. Though I am happy to hear arguments to the contrary.

feerrenrut commented on source/controlTypes/processing.py:
```diff
@@ -0,0 +1,216 @@
+from .state import State, STATES_SORTED, negativeStateLabels
+
+
+class OutputReason(Enum):
```
I'm not sure this belongs in this file. Perhaps it should be on its own?

feerrenrut commented on source/controlTypes/processing.py:
```diff
@@ -0,0 +1,216 @@
+# A part of NonVisual Desktop Access (NVDA)
```
I think this file deserves a more descriptive name.

seanbudd replied:
Do you have a suggestion? I think it's possible we make `processPositiveStates` and `processNegativeStates` private (with underscores) and rename the file `processAndLabelStates.py`. These functions appear to just be helper functions for `processAndLabelStates`. Any API breakages will have to be delayed until 2022.1, so they can be renamed in `__init__.py`

feerrenrut replied:
Yep, I'm happy with that idea.

feerrenrut reviewed (changes requested):
For now I'll assume your self review has ensured that the movement of code hasn't introduced changes. I'll first focus on the layout and introduced code.

As usual, with moved code, we'll merge rather than squash merge. Can you confirm that the commits are logical for that please. I can recommend using the  the "autosquash" feature in git. It makes it easier to keep track of what while be the final commits while making updates to them, addressing review actions etc.

seanbudd committed:
Implement Enums for Role and State constants in controlTypes (#12510)

In order to improve the NVDA API and internal typing, Enums are created
for ROLE_* and STATE_* constants.

Changes:
- retain backwards compatibility until 2022.1 in controlTypes
- improve type information for controlTypes processing
- create a clickableRoles set in role.py for processAndLabelStates
- update docstrings to reflect new types
- lint controlTypes due to the mass changes
- deprecate helper functions for processAndLabelStates
- standardise members of the State enum from 0X[0-9]+ to 0x[0-9]+
- implement a displayStringEnumMixin
- deprecate roleLabels, stateLabels, negativeStateLabels
- update coding standards for enum usage and deprecations

Co-authored-by: Leonard de Ruijter <leonardder@users.noreply.github.com>

feerrenrut commented on source/controlTypes/__init__.py:
```diff
@@ -3,207 +3,214 @@
+	INVALID_ENTRY = 0X20000
+	MODAL = 0X40000
+	AUTOCOMPLETE = 0x80000
+	MULTILINE = 0X100000
```
While here, might as well make these lower case 'x'

seanbudd commented:
I've added the following fixup commits which:

* rename `processing.py` and mark the internal helper functions as deprecated
* moves `negativeStateLabels` into a `negativeDisplayString` property. This could be followed up with a similar refactor in `braille` which could allow us to simplify the labelling process. 
* replaces `0X...` numbers with `0x...`

seanbudd commented on source/controlTypes/role.py:
```diff
@@ -0,0 +1,531 @@
+	def defaultValue(self):
+		return self.UNKNOWN
+
+	UNKNOWN = 0
+	WINDOW = 1
+	TITLEBAR = 2
+	PANE = 3
```
Thoughts on replacing this with the following?

```suggestion
	UNKNOWN = 0
	WINDOW = auto()
	TITLEBAR = auto()
	PANE = auto()
	...
```

lukaszgo1 replied:
I don't think this is a good idea as It'll break backwards compatibility for add-ons who're using integer values of these roles rather than constants. I agree this is a bad practice but I've seen this in the wild (the only example which I was able to recall quickly was @josephsl controlUsageAssistant) but there are almost certainly more. The reason this is not going to work is the fact that not all these roles are numbered sequentially for example `FOOTNOTE` is 67 and `GLASSPANE` not 68 but 69. It should be possible to implement `_generate_next_value_` for the enum and do something clever inside but IMO it is not worth the trouble.

seanbudd replied:
`WINDOW = auto()` is the equivalent of `WINDOW = 1`, however it makes it less obvious what the value is if you are using integers in a different system.

feerrenrut replied:
Yes, this is backwards compatible:
```python
>>> from enum import IntEnum
>>> class blah(IntEnum):
...     first = 0
...     second = enum.auto()
...
>>> blah.first == 0
True
>>> blah.second == 1
True
>>>
```

However, where an `enum` is matching up with some other definition, EG defined in Windows, I think it is better to be explicit.

lukaszgo1 replied:
It is backward compatible if and only if we would use `auto` only up to the `FOOTNOTE` and then  start using integers explicitly again, since as I've explained above numbering is no longer sequential from this point onward. I don't see any advantage to doing this though.

seanbudd commented on source/controlTypes/state.py:
```diff
@@ -0,0 +1,181 @@
+	@property
+	def negativeDisplayString(self) -> str:
+		"""
+		@return: The translated UI display string that should be used for when referring to this value of
+		the enum in the negative.
```
```suggestion
		@return: The translated UI display string, that should be used when referring to this enum
		member in the negative.
```

josephsl commented:
Hi,

I have largely moved away from using integers to represent rols in Control Usage Assistant. Once this PR is merged into master and during 2022.1 development cycle, I'll think about changing this add-on to using role enums (or perhaps ask the community to send in a pull request at that time).

Tahnks.

seanbudd committed:
split controlTypes.py into submodules (#12510)

The file controlTypes.py is already very complex and long, and the
code can easily be divided into 5 groups.
With the desire to group constants into Enums and maintain backwards
compatibility, which adds length and complexity, it is a canidate to be
split into submodules.

Actions:
- move controlTypes role data and functions to role.py
- move controlTypes state data and functions to state.py
- move isCurrent class to isCurrent.py
- move outputReason to outputReason.py
- move processing states functions to processAndLabelStates.py

feerrenrut reviewed (changes requested):
This is good, it's really close to being ready.

I have confirm there are only minor differences since the old HEAD ` 6ee01a53e`
I have confirmed that the moved code still matches the old code from master, except for now using Enums and other minor changes. The code seems logically equivalent

I haven't yet confirmed everything previously exposed from controlTypes.py is exposed from controlTypes package.

leonardder commented on source/controlTypes/__init__.py:
```diff
@@ -0,0 +1,245 @@
+from .state import State, STATES_SORTED, _negativeStateLabels, _stateLabels
+
+
+# After 2022, this will specify what is exported when performing a star import (from controlTypes import *).
+# Will break backwards compatibility.
+if version_year >= 2022:
+	__all__ = [
```
affect>affects?

seanbudd replied:
Thoughts on the following:

```suggestion
# To maintain backwards compatibility, all symbols are exported from the package until 2022.1
if version_year >= 2022:
	__all__ = [
	# Override (and limit) the symbols exported by the controlTypes package
	# These are the symbols available when `from controlTypes import *` is used.
```

feerrenrut replied:
Yep, reads well @seanbudd

seanbudd commented on source/controlTypes/state.py:
```diff
@@ -0,0 +1,181 @@
+		return _stateLabels
+
+	@property
+	def defaultValue(self):
+		return self.DEFUNCT
```
I was just trying to fill in default values that hopefully made some sense. On `DisplayStringEnumMixin` I can turn `defaultValue` into a `property` that returns `None` rather than an `abstractproperty` and then reraise the same `KeyError` that would be raised if there was no matching label.

feerrenrut replied:
The danger is that it leads developers to use that value whenever they aren't sure, rather than really consider what should be used. I'd like to remove the default value from the base class, it isn't clear that ALL concrete classes will have a default value and therefore breaks the "is-a" contract between base and child class. See [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle).

If there are any enums that have a confident default value, those specializations can extend the interface.

seanbudd commented on source/controlTypes/role.py:
```diff
@@ -0,0 +1,531 @@
+		return _roleLabels
+
+	@property
+	def defaultValue(self):
+		return self.UNKNOWN
```
See other comment on default values

seanbudd commented on source/controlTypes/isCurrent.py:
```diff
@@ -0,0 +1,50 @@
+	def _displayStringLabels(self):
+		return _isCurrentLabels
+
+	@property
+	def defaultValue(self):
+		return self.YES
+
```
See other comment on default values

seanbudd commented on source/controlTypes/__init__.py:
```diff
@@ -0,0 +1,245 @@
+
+from buildVersion import version_year
+
+# imports used in __all__, _ prefixed imports are marked for deprecation
```
This was more of a note for the imports below

```suggestion
# The following are imported to be exported from this folder.
# Underscore prefixed imports are marked for deprecation and can be
# removed in 2022.1
```

feerrenrut replied:
Oh I meant move all the imports and the comment inside the check. If the imports are not used, they can be automatically removed when the version ticks over.

feerrenrut commented on source/controlTypes/isCurrent.py:
```diff
@@ -0,0 +1,50 @@
+from utils.displayString import DisplayStringEnumMixin, DisplayStringEnumMixinMeta
+
+
+class IsCurrent(DisplayStringEnumMixin, str, Enum, metaclass=DisplayStringEnumMixinMeta):
```
Yes, that was what I was thinking. Expose simplified types to make then easier to use and harder to introduce subtle errors.

seanbudd commented on source/utils/displayString.py:
```diff
@@ -1,64 +1,71 @@
+
+class DisplayStringIntEnum(_DisplayStringEnumMixin, IntEnum, metaclass=_DisplayStringEnumMixinMeta):
+	"""An IntEnum class that adds a displayString property defined by _displayStringLabels"""
+	pass
```
I've since changed the line endings here to CRLF, I didn't see a need to edit earlier commits to do this as this file doesn't contain moved code.

feerrenrut replied:
When we normalize line endings, I assume we'll commit as unix line endings IE `LF. Tools like the github editor seem work with unix line endings.

feerrenrut replied:
I don't think this needs to change by the way. Just an FYI.

seanbudd replied:
Hmm, right now most are CRLF I believe.

seanbudd replied:
I prefer LF though and would make more developer environments supported I think

lukaszgo1 replied:
For what its worth  our [Coding standards](https://github.com/nvaccess/nvda/blob/master/devDocs/codingStandards.md)  currently  states:
>• Most files should contain CRLF line endings, as this is a Windows project and can't be used on Unix-like operating 
systems.

To me this reasoning sounds sane - I don't believe any half decent IDE/editor would have troubles with CRLF  whereas support for LF in Windows only development environment software cannot be taken for granted.

seanbudd replied:
Here are some issues I've run into when using VS code and CRLF when working on NVDA

* https://github.com/microsoft/vscode/issues/19254
* https://github.com/microsoft/vscode/issues/123557
* https://github.com/microsoft/vscode/issues/96104

feerrenrut replied:
Typically a mix of line endings is the real killer. Lots of tools (eg Sublime text) refuse to support it. Both because it is difficult to support, but also on philosophical grounds. Git has several options for line endings, allowing line endings to be converted on checkout / commit. I'm suggesting that we change the line endings to LF when they are committed, which is fairly standard. Much of the git ecosystem is Unix (and thus LF) oriented. When checking out code on windows, developers can have their preference of LF or CRLF.

Despite that NVDA can't be run on unix, it relies on systems that include either unix-like OS or ported unix-like tools.

feerrenrut commented on source/controlTypes/isCurrent.py:
```diff
@@ -0,0 +1,52 @@
+		try:
+			return super().displayString
+		except KeyError:
+			return self.YES.displayString
```
The previous version logged in this case: `log.debugWarning(f"No translation mapping for: {self}")`

seanbudd replied:
This should happen in the `super()` call

feerrenrut reviewed (approved):
I have confirmed (by comparing the files with `git difftool origin/master:source/controlTypes.py HEAD:source/controlTypes/__init__.py` ) that what is exposed from control types hasn't changed.

Another way to verify this may be to load use the NVDA python console:
- Import control types
- `pprint(dir(controlTypes))`.
- Save this output for both before and after this change and compare with a diff tool.

seanbudd commented:
I've updated the deprecations.md file to reflect your verification step, and changes.t2t
